src/Data/Singletons/TH/Options.hs

```diff
@@ -0,0 +1,255 @@
+    -- ** Derived functions over Options
+  , promotedTopLevelValueName
+  , promotedLetBoundValueName
+  , defunctionalizedName0
+
+    -- * OptionsMonad
+  , OptionsMonad(..), OptionsM, withOptions
+  ) where
+
+import Control.Applicative
+import Control.Monad.IO.Class (MonadIO)
+import Control.Monad.Reader (ReaderT(..), ask)
+import Control.Monad.RWS (RWST)
+import Control.Monad.State (StateT)
+import Control.Monad.Trans.Class (MonadTrans(..))
+import Control.Monad.Writer (WriterT)
+import Data.Singletons.Names ( consName, listName, nilName
+                             , mk_name_tc, mkTupleDataName, mkTupleTypeName
+                             , sconsName, sListName, snilName
+                             , splitUnderscores
+                             )
+import Data.Singletons.Util
+import Language.Haskell.TH.Desugar
+import Language.Haskell.TH.Syntax hiding (Lift(..))
+
+-- | Options that control the finer details of how @singletons@' Template
+-- Haskell machinery works.
+data Options = Options
+  { genSingKindInsts        :: Bool
+    -- ^ If 'True', then 'SingKind' instances will be generated. If 'False',
+    --   they will be omitted entirely. This can be useful in scenarios where
+    --   TH-generated 'SingKind' instances do not typecheck (for instance,
+    --   when generating singletons for GADTs).
+  , promotedClassName    :: Name -> Name
+    -- ^ Given the name of the original, unrefined class, produces the name of
+    --   the promoted equivalent of the class.
+  , promotedValueName    :: Name -> Maybe Uniq -> Name
+    -- ^ Given the name of the original, unrefined value, produces the name of
+    --   the promoted equivalent of the value. This is used for both top-level
+    --   and @let@-bound names, and the difference is encoded in the
+    --   @'Maybe' 'Uniq'@ argument. If promoting a top-level name, the argument
+    --   is 'Nothing'. If promoting a @let@-bound name, the argument is
+    --   @Just uniq@, where @uniq@ is a globally unique number that can be used
+    --   to distinguish the name from other local definitions of the same name
+    --   (e.g., if two functions both use @let x = ... in x@).
+  , singledDataTypeName  :: Name -> Name
+    -- ^ Given the name of the original, unrefined data type, produces the name
+    --   of the corresponding singleton type.
+  , singledClassName     :: Name -> Name
+    -- ^ Given the name of the original, unrefined class, produces the name of
+    --   the singled equivalent of the class.
+  , singledDataConName   :: Name -> Name
+    -- ^ Given the name of the original, unrefined data constructor, produces
+    --   the name of the corresponding singleton data constructor.
+  , singledValueName     :: Name -> Name
+    -- ^ Given the name of the original, unrefined value, produces the name of
+    --   the singled equivalent of the value.
+  , defunctionalizedName :: Name -> Int -> Name
+    -- ^ Given the original name of a function (term- or type-level) and the
```

Comment:
Do we really need this to work on both promoted and unpromoted names? That seems gratuitously unfriendly. Actually, maybe I just don't understand this documentation. It says both "original" and "type-level". How do these relate? (This is honest confusion, not just prodding.) Maybe this is because `singletons` allows the defunctionalization of user-written type families? What about type and data constructors? This documentation suggests that it works only over functions.

Reply:
> Maybe this is because singletons allows the defunctionalization of user-written type families?

Indeed, this is precisely the reason. Type synonyms can be defunctionalized too. (In fact, so can type classes, oddly enough. I've opened a separate issue about this at #429.)

> What about type and data constructors? This documentation suggests that it works only over functions.

Sort of: see #429 for the confusing state of affairs on what can be defunctionalized and what can't.

In any case, I don't think we need a precise answer to #429 yet in order to document `defunctionalizedName` better. I've revised the Haddocks for `defunctionalizedName` to remove the emphasis on functions and to explicitly mention that defunctionalization works on both term- and type-level names:

```hs
  , defunctionalizedName :: Name -> Int -> Name
    -- ^ Given the original name and the number of parameters it is applied to
    --   (the 'Int' argument), produces a type-level function name that can be
    --   partially applied when given the same number of parameters.
    --
    --   Note that defunctionalization works over both term-level names
    --   (producing symbols for the promoted name) and type-level names
    --   (producing symbols directly for the name itself). As a result, this
    --   callback is used for names in both the term and type namespaces.
```

Does this look better to you?

Reply:
Yes -- great.